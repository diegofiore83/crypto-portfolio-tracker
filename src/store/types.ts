export interface CryptoHolding {
  name: string;
  symbol: string;
  quantity: number;
  currentPrice: number;
  totalValue: number;
}
