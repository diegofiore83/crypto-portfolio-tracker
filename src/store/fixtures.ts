import { CryptoAsset, MarketData } from "./types";

export const mockCryptoAssets: CryptoAsset[] = [
  {
    id: "bitcoin",
    symbol: "btc",
    name: "Bitcoin",
    image: "https://example.com/bitcoin.png",
    current_price: 45000,
    market_cap: 850000000000,
    market_cap_rank: 1,
    total_volume: 35000000000,
    high_24h: 46000,
    low_24h: 44000,
    price_change_24h: -500,
    price_change_percentage_24h: -1.1,
    market_cap_change_24h: -10000000000,
    market_cap_change_percentage_24h: -1.2,
    circulating_supply: 19000000,
    total_supply: 21000000,
    max_supply: 21000000,
    ath: 65000,
    ath_change_percentage: -30,
    ath_date: "2021-04-14T00:00:00.000Z",
    atl: 67.81,
    atl_change_percentage: 66300,
    atl_date: "2013-07-06T00:00:00.000Z",
    roi: null,
    last_updated: "2021-09-01T00:00:00.000Z",
    fully_diluted_valuation: 850000000000,
  },
  {
    id: "ethereum",
    symbol: "eth",
    name: "Ethereum",
    image: "https://example.com/ethereum.png",
    current_price: 3000,
    market_cap: 350000000000,
    market_cap_rank: 2,
    total_volume: 20000000000,
    high_24h: 3100,
    low_24h: 2900,
    price_change_24h: -50,
    price_change_percentage_24h: -1.6,
    market_cap_change_24h: -5000000000,
    market_cap_change_percentage_24h: -1.4,
    circulating_supply: 115000000,
    total_supply: 115000000,
    max_supply: null,
    ath: 4300,
    ath_change_percentage: -30,
    ath_date: "2021-05-12T00:00:00.000Z",
    atl: 0.432979,
    atl_change_percentage: 692000,
    atl_date: "2015-10-20T00:00:00.000Z",
    roi: null,
    last_updated: "2021-09-01T00:00:00.000Z",
    fully_diluted_valuation: 350000000000,
  },
];

export const mockMarketData: MarketData = {
  prices: [
    [1738530045736, 97793.53821562871],
    [1738533735812, 96988.74415938454],
    [1738537427863, 97815.84535257479],
  ],
  market_caps: [
    [1738530045736, 1937072986217.3457],
    [1738533735812, 1921781912853.0679],
    [1738537427863, 1936077131894.8628],
  ],
  total_volumes: [
    [1738530045736, 0],
    [1738533735812, 0],
    [1738537427863, 0],
  ],
};
