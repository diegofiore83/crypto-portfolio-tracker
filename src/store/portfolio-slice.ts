import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { CryptoHolding } from "./types";

const loadFromLocalStorage = (): CryptoHolding[] => {
  const data = localStorage.getItem("portfolio");
  return data ? JSON.parse(data) : [];
};

const saveToLocalStorage = (state: CryptoHolding[]) => {
  localStorage.setItem("portfolio", JSON.stringify(state));
};

const initialState: CryptoHolding[] = loadFromLocalStorage();

const portfolioSlice = createSlice({
  name: "portfolio",
  initialState,
  reducers: {
    addCrypto: (state, action: PayloadAction<CryptoHolding>) => {
      state.push(action.payload);
      saveToLocalStorage(state);
    },
    editCrypto: (state, action: PayloadAction<CryptoHolding>) => {
      const index = state.findIndex(
        (crypto) => crypto.symbol === action.payload.symbol
      );
      if (index !== -1) {
        state[index] = action.payload;
        saveToLocalStorage(state);
      }
    },
    deleteCrypto: (state, action: PayloadAction<string>) => {
      const newState = state.filter(
        (crypto) => crypto.symbol !== action.payload
      );
      saveToLocalStorage(newState);
      return newState;
    },
  },
});

export const { addCrypto, editCrypto, deleteCrypto } = portfolioSlice.actions;
export default portfolioSlice.reducer;
