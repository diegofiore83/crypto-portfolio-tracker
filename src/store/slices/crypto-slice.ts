import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import CryptoService from "../../services/crypto-service";
import { AssetsDictionary, CryptoAsset, CryptoState } from "../types";

const initialState: CryptoState = {
  assets: {},
  status: "idle",
};

// Fetch crypto assets from the API using CryptoService
export const fetchCryptoAssets = createAsyncThunk(
  "crypto/fetchAssets",
  async () => {
    const assetArray = await CryptoService.getCryptoList();

    // Convert array to dictionary for fast lookups
    const assetsDictionary: Record<string, CryptoAsset> = assetArray.reduce(
      (acc: Record<string, CryptoAsset>, asset: CryptoAsset) => {
        acc[asset.id] = asset;
        return acc;
      },
      {} as AssetsDictionary
    );

    return assetsDictionary;
  }
);

const cryptoSlice = createSlice({
  name: "crypto",
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchCryptoAssets.pending, (state) => {
        state.status = "loading";
      })
      .addCase(fetchCryptoAssets.fulfilled, (state, action) => {
        state.status = "idle";
        state.assets = action.payload;
      })
      .addCase(fetchCryptoAssets.rejected, (state) => {
        state.status = "failed";
      });
  },
});

export default cryptoSlice.reducer;
