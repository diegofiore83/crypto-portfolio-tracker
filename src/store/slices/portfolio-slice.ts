import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { toast } from "react-toastify";
import { CryptoHolding } from "../types";

const loadFromLocalStorage = (): CryptoHolding[] => {
  const data = localStorage.getItem("portfolio");
  return data ? JSON.parse(data) : [];
};

const saveToLocalStorage = (state: CryptoHolding[]) => {
  localStorage.setItem("portfolio", JSON.stringify(state));
};

const initialState: CryptoHolding[] = loadFromLocalStorage();

const portfolioSlice = createSlice({
  name: "portfolio",
  initialState,
  reducers: {
    addCrypto: (state, action: PayloadAction<CryptoHolding>) => {
      state.push(action.payload);
      saveToLocalStorage(state);
      toast.success(
        `${action.payload.id.toUpperCase()} added to your portfolio!`
      );
    },
    editCrypto: (state, action: PayloadAction<CryptoHolding>) => {
      const index = state.findIndex(
        (crypto) => crypto.id === action.payload.id
      );
      if (index !== -1) {
        state[index] = action.payload;
        saveToLocalStorage(state);
        toast.info(
          `Updated ${action.payload.id.toUpperCase()} quantity to ${
            action.payload.quantity
          }.`
        );
      }
    },
    deleteCrypto: (state, action: PayloadAction<string>) => {
      const newState = state.filter((crypto) => crypto.id !== action.payload);
      saveToLocalStorage(newState);
      toast.error(
        `${action.payload.toUpperCase()} removed from your portfolio.`
      );
      return newState;
    },
  },
});

export const { addCrypto, editCrypto, deleteCrypto } = portfolioSlice.actions;
export default portfolioSlice.reducer;
