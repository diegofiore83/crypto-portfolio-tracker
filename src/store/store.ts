import { configureStore } from "@reduxjs/toolkit";
import cryptoReducer from "./slices/crypto-slice";
import portfolioReducer from "./slices/portfolio-slice";

const store = configureStore({
  reducer: {
    crypto: cryptoReducer,
    portfolio: portfolioReducer,
  },
});

// TypeScript types for Redux
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
