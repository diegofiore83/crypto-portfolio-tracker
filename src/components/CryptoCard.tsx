import React, { useCallback, useState } from "react";
import {
  Box,
  Card,
  CardHeader,
  CardActions,
  Grid2 as Grid,
  Button,
  TextField,
  Typography,
  Avatar,
  CardContent,
} from "@mui/material";
import { useAppDispatch } from "../store/hooks";
import { addCrypto } from "../store/slices/portfolio-slice";

interface CryptoCardProps {
  id: string;
  name: string;
  symbol: string;
  current_price: number;
  image: string;
  price_change_percentage_24h: number;
}

const CryptoCard: React.FC<CryptoCardProps> = ({
  id,
  name,
  symbol,
  current_price,
  image,
  price_change_percentage_24h,
}) => {
  const dispatch = useAppDispatch();
  const [isAddingMode, setIsAddingMode] = useState(false);
  const [quantity, setQuantity] = useState<string>("");
  const [error, setError] = useState<string | null>(null);

  const isValidQuantity = quantity !== "" && parseFloat(quantity) > 0;

  const handleQuantityChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value.trim();
    const numValue = parseFloat(value);

    setQuantity(value);
    setError(
      value === ""
        ? null
        : isNaN(numValue) || numValue <= 0
        ? "Quantity must be positive"
        : null
    );
  };

  const handleSubmit = useCallback(() => {
    if (isValidQuantity) {
      dispatch(addCrypto({ id, quantity: parseFloat(quantity) }));
      setQuantity("");
      setIsAddingMode(false);
    }
  }, [dispatch, id, quantity, isValidQuantity]);

  const handleCancel = useCallback(() => {
    setIsAddingMode(false);
    setQuantity("");
    setError(null);
  }, []);

  return (
    <Grid size={{ xs: 12, sm: 6, md: 4, lg: 3 }}>
      <Card>
        <CardHeader
          avatar={
            <Avatar
              alt={name}
              src={image}
              sx={{ width: 32, height: 32, mb: 1 }}
            />
          }
          title={
            <Box
              display="flex"
              alignItems="center"
              justifyContent="space-between"
            >
              <Typography fontWeight="bold">{symbol.toUpperCase()}</Typography>
              <Typography
                fontWeight="bold"
                color={price_change_percentage_24h > 0 ? "green" : "red"}
              >
                {price_change_percentage_24h > 0 && "+"}
                {price_change_percentage_24h.toFixed(2)}%
              </Typography>
            </Box>
          }
          subheader={
            <Typography variant="subtitle2" noWrap maxWidth="90%">
              {name}
            </Typography>
          }
        />
        {isAddingMode && (
          <CardContent>
            <Box display="flex" gap={2} alignItems="flex-start">
              <TextField
                label="Quantity"
                size="small"
                type="number"
                value={quantity}
                onKeyDown={(e) =>
                  e.key === "Enter" && isValidQuantity && handleSubmit()
                }
                onChange={handleQuantityChange}
                error={!!error}
                helperText={error}
                sx={{ flexGrow: 1, paddingBottom: error ? 0 : 6 }}
              />
              <Button
                color="primary"
                variant="contained"
                onClick={handleSubmit}
                disabled={!isValidQuantity}
              >
                Add
              </Button>
            </Box>
          </CardContent>
        )}
        <CardActions>
          <Box
            display="flex"
            justifyContent="space-between"
            alignItems="center"
            width="100%"
            paddingRight={2}
          >
            <Button
              size="small"
              color="primary"
              onClick={() =>
                isAddingMode ? handleCancel() : setIsAddingMode(true)
              }
            >
              {isAddingMode ? "Cancel" : `Add ${symbol.toUpperCase()}`}
            </Button>
            <Typography fontWeight="bold">
              $
              {current_price.toLocaleString(undefined, {
                minimumFractionDigits: current_price < 1 ? 5 : 2,
                maximumFractionDigits: 8,
                useGrouping: true,
              })}
            </Typography>
          </Box>
        </CardActions>
      </Card>
    </Grid>
  );
};

export default CryptoCard;
