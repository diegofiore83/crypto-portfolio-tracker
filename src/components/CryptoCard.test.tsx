import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import portfolioReducer from "../store/slices/portfolio-slice";
import CryptoCard from "./CryptoCard";
import { RootState } from "../store/store";
import { addCrypto } from "../store/slices/portfolio-slice";

// Mock Props
const mockProps = {
  id: "bitcoin",
  name: "Bitcoin",
  symbol: "BTC",
  current_price: 50000,
  image: "https://example.com/bitcoin.png",
  price_change_percentage_24h: 2.5,
};

describe("CryptoCard Component", () => {
  let store: ReturnType<typeof configureStore>;

  beforeEach(() => {
    store = configureStore({
      reducer: { portfolio: portfolioReducer },
      preloadedState: { portfolio: [] },
    });
  });

  const renderWithProvider = (component: React.ReactElement) =>
    render(
      <Provider store={store}>
        <MemoryRouter>{component}</MemoryRouter>
      </Provider>
    );

  test("renders CryptoCard correctly (snapshot)", () => {
    const { asFragment } = renderWithProvider(<CryptoCard {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("displays correct initial data", () => {
    renderWithProvider(<CryptoCard {...mockProps} />);

    expect(screen.getByText("Bitcoin")).toBeInTheDocument();
    expect(screen.getByText("BTC")).toBeInTheDocument();
    expect(screen.getByText("+2.50%")).toBeInTheDocument();
    expect(screen.getByText(/\$50,000(\.00)?/)).toBeInTheDocument();
  });

  test("toggles add mode when clicking the add button", () => {
    renderWithProvider(<CryptoCard {...mockProps} />);

    // Click Add Button
    fireEvent.click(screen.getByText("Add BTC"));
    expect(screen.getByLabelText("Quantity")).toBeInTheDocument(); // Input should appear

    // Click Cancel Button
    fireEvent.click(screen.getByText("Cancel"));
    expect(screen.queryByLabelText("Quantity")).not.toBeInTheDocument(); // Input should disappear
  });

  test("shows error for invalid quantity input", () => {
    renderWithProvider(<CryptoCard {...mockProps} />);
    fireEvent.click(screen.getByText("Add BTC"));

    const input = screen.getByLabelText("Quantity");

    fireEvent.change(input, { target: { value: "-1" } });
    expect(screen.getByText("Quantity must be positive")).toBeInTheDocument();

    fireEvent.change(input, { target: { value: "0" } });
    expect(screen.getByText("Quantity must be positive")).toBeInTheDocument();
  });

  test("adds crypto to portfolio when valid quantity is submitted", async () => {
    renderWithProvider(<CryptoCard {...mockProps} />);
    fireEvent.click(screen.getByText("Add BTC"));

    const input = screen.getByLabelText("Quantity");
    fireEvent.change(input, { target: { value: "2" } });

    fireEvent.click(screen.getByText("Add"));

    await waitFor(() => {
      const updatedState = (store.getState() as RootState).portfolio;
      expect(updatedState).toContainEqual({ id: "bitcoin", quantity: 2 });
    });
  });

  test("dispatches addCrypto action when submitting valid quantity", async () => {
    const dispatchSpy = jest.spyOn(store, "dispatch");

    renderWithProvider(<CryptoCard {...mockProps} />);
    fireEvent.click(screen.getByText("Add BTC"));

    fireEvent.change(screen.getByLabelText("Quantity"), {
      target: { value: "3" },
    });
    fireEvent.click(screen.getByText("Add"));

    await waitFor(() => {
      expect(dispatchSpy).toHaveBeenCalledWith(
        addCrypto({ id: "bitcoin", quantity: 3 })
      );
    });
  });
});
