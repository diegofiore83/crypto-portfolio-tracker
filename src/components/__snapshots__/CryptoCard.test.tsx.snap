// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoCard Component renders CryptoCard correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-sm-6 MuiGrid2-grid-md-4 MuiGrid2-grid-lg-3 css-bz5kom-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <a
        class="MuiCardHeader-root css-2vrjko-MuiCardHeader-root"
        data-discover="true"
        href="/crypto/bitcoin"
      >
        <div
          class="MuiCardHeader-avatar css-1r9wl67-MuiCardHeader-avatar"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-ci7oj7-MuiAvatar-root"
          >
            <img
              alt="Bitcoin"
              class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
              src="https://example.com/bitcoin.png"
            />
          </div>
        </div>
        <div
          class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-7ctihp-MuiTypography-root"
          >
            <div
              class="MuiBox-root css-1lekzkb"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-lhsbas-MuiTypography-root"
              >
                BTC
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-pvlz4o-MuiTypography-root"
              >
                +2.50%
              </p>
            </div>
          </span>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-noWrap css-14ggmo7-MuiTypography-root"
          >
            Bitcoin
          </h6>
        </div>
      </a>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
      >
        <div
          class="MuiBox-root css-1kms82h"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-1ccsd5i-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Add BTC
          </button>
          <p
            class="MuiTypography-root MuiTypography-body1 css-lhsbas-MuiTypography-root"
          >
            $50,000.00
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
